src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj

```diff
@@ -58,6 +58,9 @@
     <ExpectedDependency Include="System.Runtime.CompilerServices.Unsafe" OptimizeAssemblies="lib/netstandard2.0/System.Runtime.CompilerServices.Unsafe.dll" />
     <ExpectedDependency Include="System.Text.Encoding.CodePages" OptimizeAssemblies="lib/netstandard2.0/System.Text.Encoding.CodePages.dll" />
     <ExpectedDependency Include="System.Numerics.Vectors"/>
+    <ExpectedDependency Include="Nerdbank.Streams"/>
```

Comment:
new dependency from StreamJsonRpc 2.x tagging @tmat

Reply:
Is Nerdback.Streams already installed by platform? If that's the case then it should be listed in the list below `<_Dependency Remove=.../>` along with StreamJsonRpc, Newtonsoft.Json etc.

---
In reply to: [263634732](https://github.com/dotnet/roslyn/pull/33960#discussion_r263634732) [](ancestors = 263634732)

Reply:
Yes: StreamJsonRpc 2.0 and Nerdbank.Streams is already included in VS.

Reply:
Which TFM? netstandard2.0 or net46?

---
In reply to: [263945370](https://github.com/dotnet/roslyn/pull/33960#discussion_r263945370) [](ancestors = 263945370)

Reply:
[The build we insert into VS](https://dev.azure.com/devdiv/DevDiv/_git/VS?path=%2Fsrc%2FSetupPackages%2FCoreIde%2FConnected%2Fcore%2Ffiles.swr&version=GBlab%2Fd16.0stg&line=47&lineStyle=plain&lineEnd=48&lineStartColumn=1&lineEndColumn=1) is for `net46`.

Reply:
[Nerdbank.Streams](https://www.nuget.org/packages/nerdbank.streams) has other TFMs in the package including netstandard2.0.

Reply:
So the reason why `System.IO.Pipelines` and `System.ValueTuple` show up is that the `net46` version of Nerdbank.Streams is picked up and it needs `System.IO.Pipelines` and `System.ValueTuple`. NuGet always prefers the same framework, so that's why it's not picking up netstandard2.0 from net472 project.

So, for now we should move `System.IO.Pipelines` and `System.ValueTuple` to the list below where we remove dependencies that are already in VS.

@AArnott Any reason not to insert `netstandard2.0` into VS instead of `net46`?

Reply:
The netstandard2.0 build of the package has the same dependencies, if you look at its [metadata on nuget.org](https://www.nuget.org/packages/nerdbank.streams).

> Any reason not to insert netstandard2.0 into VS instead of net46?

Since NuGet always considers a concrete runtime as preferable to netstandard, that's how I design my packages. In this case that suggests `net46` *is* preferable over `netstandard2.0` when we're on .NET Framework (probably because .NET Framework offers APIs we use that netstandard2.0 lacks). We might add a net472 target if that would somehow help this situation. But it's not clear to me how it would. Thoughts?

Reply:
@heejaechang Please add a comment to `System.ValueTuple` that it is included due to `Nerdbank.Streams` targeting `net46`.

Reply:
@AArnott Adding `net472` target to the package would remove the reference to `System.ValueTuple` facade. That one is not referenced by `netstandard2.0` dll (check the metadata references, not package references). 

Another benefit would be that VS components that target `netstandard2.0` would be able to use `StreamJsonRpc`. They can do that currently, but there is a potential danger that they will use an API that is in netstandard2.0 but not in net46 and then be broken in VS.

Reply:
With https://github.com/Microsoft/vs-streamjsonrpc/pull/248, there is no functionality loss in StreamJsonRpc between the net46 and netstandard2.0 builds. But before we switch VS to the netstandard2.0 build, I'd like to know what benefit it brings, and whether we can achieve the same by targeting net472 (so that VS consumes what NuGet would have had it consume if it was resolving the package).

Reply:
> With Microsoft/vs-streamjsonrpc#248, there is no functionality loss in StreamJsonRpc between the net46 and netstandard2.0 builds

I assumed so, that's why I said _potential_. 

> whether we can achieve the same by targeting net472

Yes, net472 and netstandard2.0 can be the same binary.

> so that VS consumes what NuGet would have had it consume if it was resolving the package

That depends what the project that references the package targets. If it targets netstandard2.0 then nuget picks netstandard2.0. But if we add a copy of the netstandard2.0 binary to net472 it wouldn't matter whether the project targets net472 or netstandard2.0

Reply:
BTW, is StreamJsonRpc only used in VS, or is it more generally useful library? If it's VS only then targeting anything but netstandard2.0 is not necessary.

Reply:
> Adding net472 target to the package would remove the reference to System.ValueTuple facade

True. But it doesn't (naturally) remove the package reference. So would this actually solve a problem for you? Is there an RPS concern from loading this facade assembly (i.e. are we _really_ close to avoiding all such references)?

> there is a potential danger that they will use an API that is in netstandard2.0 but not in net46 and then be broken in VS.

No danger. We never define APIs in netstandard that are missing from the concrete platform targets. netstandard is always a strict subset of the concrete platforms for exactly this reason. (and Nuget seems to expect this policy as well).

> net472 and netstandard2.0 can be the same binary.

Does this require custom build authoring to manually copy the netstandard2.0 build into the `lib\net472` folder of the nuget package? Or is this properly supported another way? Otherwise I'd tend to just add `net472` to the `TargetFrameworks` property (which slows the build down a bit and I need to run tests against a new build, but I'm used to that).

> That depends what the project that references the package targets. 

VS targets .NET Framework, so ultimately the runtime assembly NuGet and the build _would_ pick (if we built VS with the features we ship) would be an assembly from a `net4xx` folder. But ya, we totally design our packages so that any individual project can safely reference a netstandard build of our assembly yet run against a .NET Framework build of the assembly. So I'm comfortable with that.

So I think we're fine here already. And it's not clear to me (unless you have some really fancy build authoring) that simply removing an assembly reference to a facade assembly will solve a problem for you. But if you think it will actually avoid an assembly load of a facade assembly at runtime in VS, that's worthwhile.

Reply:
> is StreamJsonRpc only used in VS, or is it more generally useful library? If it's VS only then targeting anything but netstandard2.0 is not necessary.

Yes, StreamJsonRpc is a general-purpose JSON-RPC library which is used in many places already.

Reply:
> Does this require custom build authoring to manually copy the netstandard2.0 build into the lib\net472 folder of the nuget package? 

A few lines of msbuild xml would do. Something similar to
https://github.com/dotnet/roslyn/blob/master/src/NuGet/Microsoft.Net.Compilers/Microsoft.Net.Compilers.Package.csproj#L18
https://github.com/dotnet/roslyn/blob/master/src/NuGet/Microsoft.Net.Compilers/Microsoft.Net.Compilers.Package.csproj#L29-L37

> But if you think it will actually avoid an assembly load of a facade assembly at runtime in VS, that's worthwhile.

It will avoid the load if there is any method in StreamJsonRpc that uses ValueTuple and is not NGEN'd.